### DISCUSSION TITLE
Android service stops in background even if battery optimisations off

### DISCUSSION THREAD

**Discussion by TtuxX** [rustdesk/rustdesk#5627] [General]
### Bug Description

Hi there ! :)

First a huge thank you for having developed and maintaining your incredible app and all the different platforms client while providing it open source and giving the selfhosting possibility!!

Back to the actual issue I am experiencing: I have 2 android devices (NVIDIA Shield TV Pro 2019 and Pixel phone on GrapheneOS), and on both devices I cannot connect anymore in background after some time, even if I have disabled battery optimisations and even if I have not cleared the app from recent apps.

Do you know if there is a specific reason for it ? It appears android client version 1.2.3 does not has the notification permission as it does not emit notifications. 
Would it be possible to add a persistent notification when a session is launched to further prevent the app from being unreachable when in background position? 

### How to Reproduce

- Install android client 1.2.3
- Launch the app and create a permanent password while disabling battery optimisations and toogling on the startup launch button
- Connect to the android app from my desktop
- Switch to other apps on the phone and do other things on the phone from the desktop
- At some moments the android phone network connection drops
- Both devices usually reconnect up to the point where at a moment it does not work
- Try to connect all over again (having hoped that the app is still active in background)
- The app is not reachable anymore from desktop
- Go back to the app from your phone and see that service has been deactivated

### Expected Behavior

Keep the app active in background (through a permanent notification if needed).

### Operating system(s) on local side and remote side

Windows 11 -> Android 13 (latest GrapheneOS release)

### RustDesk Version(s) on local side and remote side

1.2.2 -> 1.2.3

### Screenshots

Nothing really specific

**Comment by rustdesk** [comment]
> Would it be possible to add a persistent notification when a session is launched to further prevent the app from being unreachable when in background position?

Frankly, we do not have a lot of experience on this.

**Comment by rustdesk** [comment]
> Keep the app active in background (through a permanent notification if needed).

any example?

**Reply by TtuxX** [reply]
Yes I do have some examples in mind :)

For example, this is what I have with Adguard:

<img src="https://github.com/rustdesk/rustdesk/assets/6570830/37cfcaa0-4855-4b3f-9cb4-2ff8b95eebf7" width=70% height=70%>

If we derive the idea from this, we could maybe imagine something a bit like that (the notification would be silent and folded by default but on Android you could unfold it using the button on top right of the notification):

<img src="https://github.com/rustdesk/rustdesk/assets/6570830/6ba5250e-dbd9-4d63-b83d-59eed7ee8609" width=70% height=70%>

**Reply by funtu729183-lab** [reply]
After every single setting enable whenever someone lock the phone it's Stop working and screen go blank and if I connected again it's Will showing me waiting for the image..

PHONE REDMI 13 5G HYPER OS
PLEASE HELP ME 🙏

**Comment by rustdesk** [comment]
Implement an Android floating window to prevent the system from killing the service, ready in nightly build.
 https://x.com/rustdesk/status/1798387103862231205

**Reply by TtuxX** [reply]
Awesome, thank you so much for your time, work and for considering and addressing this issue.
You are the best! 🎉

**Reply by TtuxX** [reply]
@rustdesk @21pages Actually now I have an issue: on my Android TV, I always see the new overlay button, even when watching movies in full-screen and so on...
Would it be possible to add a transparency setting for this button that could go as high as 100% transparency? And could the button position be saved upon application relaunch? Thank you again a lot!! 🙏🏻

**Reply by rpm5099** [reply]
> Implement an Android floating window to prevent the system from killing the service, ready in nightly build. https://x.com/rustdesk/status/1798387103862231205

I'm not quite sure why, but even with the floating window enabled, battery unoptimized, and every permission asked for granted including accessibility, the service continually gets killed.  Usually on reboot it starts again, so it's not a matter of permission at startup, something is actively killing the background service.  I am having this problem on 3 different android phones.  Is there any way of retrieving logs I can send?  I really like rustdesk over a number of other products I have tried, paid and free, but unfortunately without reliable unattended full remote access it's not going to work.

Even a tasker or automate job that just starts the service every 5 minutes if it's not running would be workable.  I do have some devices rooted, but without TWRP (which isn't supported on some of my devices) I have to worry about Google screwing me with an undesired update and unrooting, which they have done a number of times.  

Really frustrating - Google GTF OFF my devices.  I'm starting to think they are almost as bad as Apple on mobile devices.

**Comment by rpm5099** [comment]
I am using the floating feature and the service still stops multiple times a day.

Perhaps Google should understand that we own these devices and stop interfering with how we use them.

**Comment by cruisermaxx** [comment]
I also find that the connection can't be maintained from Android control of Rustdesk.
It's weird. Is it that hard to stay alive in background on Android?

**Reply by funtu729183-lab** [reply]
After every single setting enable whenever someone lock the phone it's Stop working and screen go blank and if I connected again it's Will showing me waiting for the image..

PHONE REDMI 13 5G HYPER OS
PLEASE HELP ME 🙏

**Comment by cruisermaxx** [comment]
According to the findings, the issue is mainly caused by network changes — for example, switching from a cellular network to Wi-Fi (or vice versa), establishing a wireless ADB connection, or enabling/disabling a VPN. It is unclear whether there is a way to automatically re-establish the connection in the background.

**Reply by rpm5099** [reply]
What findings are you referring to?  My devices have a permanent always on VPN, but obviously do have to switch from cellular to wifi.  If the android kernel has some feature that screws with services and permissions based on switching data networks I would really like to find it because I'll go the ends of the earth to disable it.

**Comment by theunderpantcoder** [comment]
Still a problem in 2026. I connect from a linux desktop, or a windows laptop with the same result. Various Samsung Android phones running 13 or 15. The Rustdesk service cannot be relied upon for continuous operation on Android. I even tried with a WIFI connected phone with no sim card. Worked for 3 days then the service died for no apparent reason. However it is only the screen casting which is dead, the phone can still be connected to, but Rustdesk just displays a 'Waiting for Screen' message. The "Mobile Actions" button does nothing so the Rustdesk sits there with a blank screen. Nothing can be done. If the service is restarted on the phone Rustdesk does not auto-reconnect, and a new connection must be issued from the host. Whan it works Rustdesk is great, but it's unreliable. So I've given up with my plan to leave the phone at home and connect remotely while I'm abroad.

**Reply by funtu729183-lab** [reply]
After every single setting enable whenever someone lock the target phone it's Stop working and screen go blank and if I connected again it's Will showing me waiting for the image..

PHONE REDMI 13 5G HYPER OS
PLEASE HELP ME 🙏
